### PR TITLE
Extended BeautifulSoup Reader loader.

### PR DESCRIPTION
**Pull request by Smyja**
This Pull request enables the reading of an entire documentation and not just a single page. I have also added a readme.com reader for readme.com documentations.

### CODE REVIEW

File: loader_hub/web/beautiful_soup_web/base.py [hunk 1]
```diff
@@ -33,9 +84,9 @@ class BeautifulSoupWebReader(BaseReader):
 
     Reads pages from the web.
     Requires the `bs4` and `urllib` packages.
-
+    
     Args:
-        file_extractor (Optional[Dict[str, Callable]]): A mapping of website
+        website_extractor (Optional[Dict[str, Callable]]): A mapping of website
```
Comment (emptycrown) [inline loader_hub/web/beautiful_soup_web/base.py:87]: Good catch

File: loader_hub/web/beautiful_soup_web/base.py [hunk 2]
```diff
@@ -15,16 +15,67 @@ def _substack_reader(soup: Any) -> Tuple[str, Dict[str, Any]]:
...
+            
+def _readmedocs_reader(soup: Any,url: str) -> Tuple[str, Dict[str, Any]]:
+    """Extract text from a ReadMe documentation site"""
+    
+    links = soup.find_all("a")
+    docs_links = [link["href"] for link in links if "/docs/" in link["href"]]
+    docs_links = list(set(docs_links))
+    for i in range(len(docs_links)):
+        if not docs_links[i].startswith("http"):
+            docs_links[i] = url + docs_links[i]
```
Comment (emptycrown) [inline loader_hub/web/beautiful_soup_web/base.py:54]: Same as above. I think `urljoin` should be used.

File: loader_hub/web/beautiful_soup_web/base.py [hunk 3]
```diff
@@ -15,16 +15,67 @@ def _substack_reader(soup: Any) -> Tuple[str, Dict[str, Any]]:
...
+    for link in links:
+        rtd_links.append(link['href'])   
+    for i in range(len(rtd_links)):
+        if not rtd_links[i].startswith("http"):
+            rtd_links[i] = url + rtd_links[i]
+    texts = []
+    broken_docs = []
+    
+    for doc_link in rtd_links:
+        import requests
```
Comment (emptycrown) [inline loader_hub/web/beautiful_soup_web/base.py:33]: These can be pulled out of the for loop and just put at the top of this function

File: loader_hub/web/beautiful_soup_web/base.py [hunk 4]
```diff
@@ -15,16 +15,67 @@ def _substack_reader(soup: Any) -> Tuple[str, Dict[str, Any]]:
...
+        except IndexError:
+            text = None
+            raise IndexError("No content found")
+
+        if text:
+            texts.append("\n".join([t for t in text.split("\n") if t]))
+    return '\n'.join(texts), {}
 
 
 DEFAULT_WEBSITE_EXTRACTOR: Dict[str, Callable[[Any], Tuple[str, Dict[str, Any]]]] = {
```
Comment (emptycrown) [inline loader_hub/web/beautiful_soup_web/base.py:73]: If `url` is now an arg, the type here is now `Dict[str, Callable[[Any, str], Tuple[str, Dict[str, Any]]]] `

**Review by emptycrown — CHANGES_REQUESTED** [review]
Thanks for doing this! This definitely makes the documentation website parsers a lot better. Left a few comments.

**Comment by Smyja** [comment]
> Thanks for doing this! This definitely makes the documentation website parsers a lot better. Left a few comments.

Thanks for the review! I have made the changes.

File: loader_hub/web/beautiful_soup_web/base.py [hunk 5]
```diff
@@ -80,11 +131,11 @@ def load_data(
             data = ""
             extra_info = {"URL": url}
             if hostname in self.website_extractor:
-                data, metadata = self.website_extractor[hostname](soup)
-                extra_info.update(metadata)
+                data = self.website_extractor[hostname](soup,url)
```
[the remark this answers is not in the document]
Comment (Smyja) [reply loader_hub/web/beautiful_soup_web/base.py]: I have added it back.

File: loader_hub/web/beautiful_soup_web/base.py [hunk 6]
```diff
@@ -15,16 +15,67 @@ def _substack_reader(soup: Any) -> Tuple[str, Dict[str, Any]]:
...
-    return "\n".join([t for t in text.split("\n") if t]), {}
+    """
+    links = soup.find_all('a', {'class': 'reference internal'})
+    rtd_links = []
+    
+    for link in links:
+        rtd_links.append(link['href'])   
+    for i in range(len(rtd_links)):
+        if not rtd_links[i].startswith("http"):
+            rtd_links[i] = url + rtd_links[i]
```
[the remark this answers is not in the document]
Comment (Smyja) [reply loader_hub/web/beautiful_soup_web/base.py]: Updated this.

File: loader_hub/web/beautiful_soup_web/base.py (hunk 2 above)
Comment (Smyja) [reply loader_hub/web/beautiful_soup_web/base.py]: updated this.

**Review by emptycrown — APPROVED** [review]

**Pull request merged**